### ISSUE TITLE
kubectl port-forward container listener to local port

### ISSUE THREAD

**Issue by mdaniel** [kubernetes/kubernetes#20227]
It would be great if `kubectl port-forward` offered the equivalent of `ssh -R`, such that it would open a _listening_ port on the container that forwards to a port on the local machine. I would be perfectly content if that port were only available to the container (i.e. just the 3-tuple `-R` flavor, without `bind_address`; no need to have other containers be able to connect to the port)

Since there is no `kubectl cp` command that I could find, this would make for a painless way to get content into a running container via http from a workstation

**Comment by bgrant0607** [comment]
kubectl cp is #13776

cc @ncdc

**Comment by spiffxp** [comment]
/sig cli
If the original intent of this issue is satisfied by `kubectl cp` this can probably be closed, since that's been implemented

**Comment by witrin** [comment]
@spiffxp What about debugging a running container? With a simple line to create a reverse tunnel to the container would be very helpful!

**Comment by witrin** [comment]
Port-forwarding in one direction is easy and helpful for debugging `kubectl port-forward <pod> 80:80` but the inverse is missing (e.g. `kubectl port-forward --reverse <pod> 9000:9000`). The simplest but also insecure way to solve this problem in my case was to deploy a [CronJob DaemonSet](https://github.com/kubernetes/kubernetes/issues/36601) which allows `GatewayPorts` on the Host. Maybe there are better workarounds but the task is pretty common I think. Alternatives would include proxy in a container but that would add more costs as needed I think.

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-cli-feature-requests

**Comment by shiywang** [comment]
cc @vfreex you might be interested

**Comment by vfreex** [comment]
It would be nice to have this feature. However, like @witrin said, it is hard to implement.

In order to capture the traffic targeting the remote port, a program running in Pod is required to listen on that port. In the case of `ssh -R`, `sshd` running on the remote side will open a port. But in the case of a Pod, I don't think it is a good idea to put an agent like `sshd` to do so.

Any ideas?

**Comment by ncdc** [comment]
I would think this would just be modifying the arguments passed to `socat` in the kubelet/cri. It should be possible to do.

**Comment by witrin** [comment]
You doing this with `socat`? :grin: That's funny, because I thought about building a container running `socat` to get this done, so I have not to touch the configuration of the hosts SSH daemon. If this is really cheap to implement like @ncdc estimated it would be such a helpful feature.

**Comment by ncdc** [comment]
Yes, `kubectl port-forward` uses `socat` in the kubelet/cri. This would be a matter of modifying the protocol to be able to specify remote vs local ports, and then adding the support for the other direction.

**Comment by vfreex** [comment]
@ncdc That would be very interesting. I would like to take a look.

**Comment by witrin** [comment]
@vfreex How could I test this when I'm using GKE?

**Comment by vfreex** [comment]
@witrin, do you mean testing my PR? If so, I will update my PR a few days later, after the new year's day. The functionality has been almost completed.

**Comment by drewwells** [comment]
I see a PR, is there a workaround available to do this with a socat sidecar container?

**Comment by bbzg** [comment]
Any progress on this? Would be awesome 🚀

**Comment by vfreex** [comment]
Hi @bbzg,
This PR is actually a PoC, which is already working. However, the Kubernetes community refuses to add new features until SPDY is deprecated and migrated to WebSocket or HTTP/2.

**Comment by dims** [comment]
@vfreex i see a PR in k/k - https://github.com/kubernetes/kubernetes/pull/57320 - that does not seem to have objections from anyone. Though i do see that it's terribly out of date. Please rebase to latest master and let's try getting a formal objection from someone in sig-networking.

**Comment by vfreex** [comment]
@dims Thank you. Sorry for my late update because I was so busy recently. I've rebased the PR so that it can be tried with the latest version.
Since it involves many SIGs, it could not be easy to get approved. I am planning to open a KEP to track this issue. Do you have any suggestions or anyone I can ask for help?

**Comment by AceHack** [comment]
Any update or workaround?  I would really like this feature.

**Comment by aurolle** [comment]
Hi, I really need these feature. Is there any update?

**Comment by tomekit** [comment]
Bumping to prevent stale. Is there any update here?

**Comment by excavador** [comment]
Any updates?

**Comment by tomekit** [comment]
Not a solution really, but I've temporarily solved my problem using: `ngrok.io` which is 3rd party service exposing your local ports to public internet. Not ideal from many perspectives including, latency, security (do you trust 3rd party?).
Something that many won't be able to accept as a production solution, but this satisfies my needs to debug things on remote environment (very rarely done anyway).

I still hope there will be an official solution that can address this issues natively.

**Comment by AceHack** [comment]
Bump

**Comment by avik-so** [comment]
@vfreex Has SPDY been deprecated and migrated to WebSocket or HTTP/2 yet?

**Comment by glensc** [comment]
@tomekit (https://github.com/kubernetes/kubernetes/issues/20227#issuecomment-469010698): can you post your `ngrok` example details here?

**Comment by AceHack** [comment]
Any update?

**Comment by ikogan** [comment]
We're encountering this situation when trying to debug with PyCharm. Since PyCharm expects the app [being debugged to connect to it](https://www.jetbrains.com/help/idea/run-debug-configuration-python-remote-debug.html), we would need to find a way to forward traffic from the container to the local machine. Is there any update on this?

**Comment by nathan-beneke** [comment]
I would also really appreciate this feature

**Comment by guusvw** [comment]
Bump!

**Comment by avik-so** [comment]
@nathan-beneke  @guusvw Have you tried Telepresence? https://www.telepresence.io/

**Comment by ikogan** [comment]
While Telepresence technically does the job, it doesn't really integrate with many of the tools like [skaffold](https://github.com/GoogleContainerTools/skaffold) and PyCharm.

**Comment by excavador** [comment]
> @nathan-beneke @guusvw Have you tried Telepresence? https://www.telepresence.io/

Telepresence does this job terrible
- it modifies you kubernetes state, replace your objects by own objects. If you lost connection your installation would be broken
- it does not allow to port-forward only single port, only all ports of deployment/pod

More precisely, I want to see solution which intercept network connection initiation to specific service/deployment/pod port and forward it to my local tool.
If kubectl lost connection to cluster, all incoming network connections to this port should go in previous way (without any changes in cluster)

Telepresence definitely does not address these issues.

**Comment by krmayankk** [comment]
Tech-debt: (Sig-Architecture)
Category: Usability
Reason:   Improve debuggability for editors, help attach debugger to containers

**Comment by ColinChartier** [comment]
I wrote an article on how to do this with only FOSS tools: https://layerci.com/blog/container-tcp-tunnel/

**Comment by excavador** [comment]
> I wrote an article on how to do this with only FOSS tools: https://layerci.com/blog/container-tcp-tunnel/
> 
> TL;DR: `nc 127.0.0.1 8000 | kubectl exec -i web-pod tcpserver 127.0.0.1 8080 cat`

You solution does not work in case of https://github.com/GoogleContainerTools/distroless containers or in case if `nc` tool missed inside container.

The generic solution for any possible docker images possible only on kubernetes level.

**Comment by ColinChartier** [comment]
> > I wrote an article on how to do this with only FOSS tools: https://layerci.com/blog/container-tcp-tunnel/
> > TL;DR: `nc 127.0.0.1 8000 | kubectl exec -i web-pod tcpserver 127.0.0.1 8080 cat`
> 
> You solution does not work in case of https://github.com/GoogleContainerTools/distroless containers or in case if `nc` tool missed inside container.
> 
> The generic solution for any possible docker images possible only on kubernetes level.

Yes, but it's possible to build a statically linked "tcpserver" and copy it in if necessary - which should work for almost all use cases

**Comment by excavador** [comment]
> Yes, but it's possible to build a statically linked "tcpserver" and copy it in if necessary - which should work for almost all use cases

Of course, it possible. You have the number of options:
- telepresence
- include nc to your base image or some your tool
- write the light version of telepresence

In my company, we first write telepresense equivalent, now returns redirects from our nginx-based container and process it on our side.

But ticket is not about workarounds, ticket about built-in universal solution to address reverse port-forward issues.

**Comment by GuyPaddock** [comment]
@ColinChartier I found the solution you proposed to really only work in a single-shot fashion, for sending back a single response. It doesn't work well for a PyCharm/PHPStorm/XDebug scenario where: 1) client and server need to communicate back and forth during a single connection, and 2) the remote end needs to be listening across multiple remote requests.

Instead, here's a solution that involves spinning-up a sidecar SSH container + service into the cluster that just has SSH forward a remote port locally:
https://github.com/GuyPaddock/inveniem-nextcloud-azure/blob/develop/launch_xdebug_proxy.sh

@ikogan Let me know if this type of solution would work for your PyCharm scenario. I wrote the script for PHP XDebug but there's nothing protocol-specific in the script.

Prior to my SSH-based approach, I tried crafting a solution with `socat` running remotely that communicates with a `socat` instance running locally, but it ended up really convoluted because:
1. The remote instance had to have two listeners -- one to accept incoming XDebug connections from PHP, and one to accept a connection from the `socat` instance running on the developer machine.
2. The local instance had to have two clients -- one to connect to the remote `socat` instance (over a `kubectl port-forward`, and one to connect to the IDE.
3. Keeping this going required having both the local end and remote end fork on connection, since otherwise `socat` would quit after a single connection. But forking on the local end resulted in the equivalent to a busy-poll fork bomb because it would just keep trying to connect to the remote end.

For those curious, here's how this bad idea looked:
https://gist.github.com/GuyPaddock/783957b1b6d7d89751d1796a2448ce5d

**Comment by omrikiei** [comment]
Hey, until there's a native solution out there, here's ktunnel - https://github.com/omrikiei/ktunnel

**Comment by trinhpham** [comment]
/remove-lifecycle stale

Still looking for an update.

**Comment by Eriner** [comment]
I released an open source kubectl plugin to proxy Kubernetes Services on Thursday: https://github.com/soluble-ai/kubetap

I am anxiously awaiting movement on this issue, or in @vfreex's [yet unmerged PR](https://github.com/kubernetes/kubernetes/pull/57320) which has been on hold for more than a year:

@smarterclayton
>@mikedanese and I have been talking about security and safety related changes to how kubelet to apiserver communication works in light of last months security vulnerability. Until we have some time to discuss how that might evolve, we want to limit the number of changes made to the related protocols.
>
>For that reason I’m going to put a hold on this - we may decide to remove Spdy altogether and this needs to be a part of that usecase.

Is [this](https://github.com/kubernetes/kubernetes/pull/57320#issuecomment-451626316) still a reason to not pursue this **extremely useful** feature? I am concerned by not merging this after four years, it is encouraging people (like me) to jump through hoops... that are on fire... over lava.

**Comment by vitaliitylyk** [comment]
This feature would have been extremely useful for all sorts of local development scenarios. I really hope this gets implemented.

**Comment by EtnaRyanK** [comment]
+1

**Comment by lifesboy** [comment]
+1

**Comment by excavador** [comment]
+1

**Comment by AkihiroSuda** [comment]
My this project supports container-to-local port forwarding as well as local-to-container and container-to-container: https://github.com/norouter/norouter

**Comment by rikatz** [comment]
Hi :) 

Does the kubectl debug command (https://kubernetes.io/docs/tasks/debug-application-cluster/debug-running-pod/#ephemeral-container) with ephemeral containers solve this?

If so, this issue can be closed as the feature is already in Alpha, right?

cc @verb

**Comment by rikatz** [comment]
An additional thing, I understand that the idea is to open a local port that forwards the content to a side container, as an example so I actually don't know if you can create an ephemeral container and do a Port Forward into that.

Let me know if there's something else I'm missing here :)

**Comment by Eriner** [comment]
> Does the kubectl debug command (https://kubernetes.io/docs/tasks/debug-application-cluster/debug-running-pod/#ephemeral-container) with ephemeral containers solve this?

No.

/remove-lifecycle stale

**Comment by davinchia** [comment]
+1 to this. We have Java processes backed by Kube pods with sidecars that pipe stdio to the original calling process. This works fine if the calling process is in Kube. This does not work if the callling process isn't in Kube space. Having reverse port forward would allow us to test these Kube-backed processes as part of the standard Java test suite, instead of jumping through manual hoops.

**Comment by rymnc** [comment]
Having a kubectl native method to support remote port forwarding would be great

**Comment by atorr0** [comment]
There's a `kubectl ssh` plugin that I think can be used with the -R (though I didn't tried it) at https://github.com/jordanwilson230/kubectl-plugins

**Comment by quantommy** [comment]
Out of curiosity did someone use this project for reverse tunneling? https://github.com/omrikiei/ktunnel

**Comment by omrikiei** [comment]
@tomekit well, I obviously did :)
my work has been using it for a while with tilt.dev to add remote debugging to pycharm...
feel free to reach out to me if you have any questions

**Comment by moshebe** [comment]
This tool seems to solve the issue for me as well:
https://github.com/omrikiei/ktunnel
🎉

**Comment by rusanki** [comment]
would be good to have this reverse proxy setup for local development and debugging

**Comment by svenkroll** [comment]
Ktunnel rescued my life! Won't fight with telepresence or similar, please add the ktunnel functionality to kubectl.

**Comment by Eriner** [comment]
> [mdaniel](https://github.com/mdaniel) opened this issue on Jan 27, 2016 · 86 comments

In less than a year and a half, I look forward to celebrating this feature request having been open for a decade.

I've long since left k8s due to a lack of basic tooling like this, but I've grown an odd attachment to this thread. Seeing the consistent bumps via the "/remove-lifecycle stale" posts always leaves me with a feeling of nostalgia. Every so often when I see the bump, I wonder how @mdaniel and @vfreex must feel, if they haven't muted the thread long ago.

At one time I wanted k8s to get this feature, but after so many years I think I'd miss seeing this thread rise to the top of my inbox every so often.

**Comment by phoenix-bjoern** [comment]
IMHO we should keep this issue open. It would be a great improvement allowing many new use cases.

**Comment by zeppelinen** [comment]
Almost 10 years

**Comment by exxocism** [comment]
Please
